### ISSUE TITLE
ui-router with html5mode(true) page refresh issue

### ISSUE THREAD

**Issue by reggieboyYEAH** [angular-ui/ui-router#1985]
I have an app that uses **angular-ui-router** with **html5mode(true)**. Everything seems to work fine when running and routing to other states.

My default state is _app/calendar_ which is set during module.run()

But when i refresh the page while i'm currently in other routes(lets say _app/profile_) it takes me back to app/calendar.

Debugging i noticed that the $state.current is always empty after i refresh the page

```
Object {name: "", url: "^", views: null, abstract: true}
```

if only the $state.current has value i can just transistion to the current state.

Is there anything that i am missing?

Hopefully someone can help.

My **server routing** looks like

```
app.get('/:var(/|app/calendar|app/customers|app/profile|app/settings)?', function(req, res) {
    res.sendFile('/app/main/main.html',{ root: '../Appt/public' });
});
```

i'm always serving the same file.

and my **front-end state configuration** 

```
(
    function()
    {
        angular.module('Appt.Main').config(['$stateProvider','$locationProvider',function($stateProvider,$locationProvider)
        {
            $locationProvider.html5Mode(true);

            var calendar = {
                    name: 'calendar',
                    url: 'app/calendar',
                    controller: 'Appt.Main.CalendarController',
                    controllerAs: 'calendar',
                    templateUrl: '/app/main/calendar/calendar.html'
                },
                customers = {
                    name: 'customers',
                    url: 'app/customers',
                    controller : 'Appt.Main.CustomersController',
                    controllerAs : 'customers',
                    templateUrl : '/app/main/customers/customers.html'
                },
                profile = {
                    name: 'profile',
                    url: 'app/profile',
                    controller : 'Appt.Main.ProfileController',
                    controllerAs : 'profile',
                    templateUrl : '/app/main/profile/profile.html'
                },
                settings = {
                    name: 'settings',
                    url: 'app/settings',
                    controller : 'Appt.Main.SettingsController',
                    controllerAs : 'settings',
                    templateUrl : '/app/main/settings/settings.html'
                };


            $stateProvider.state(calendar);
            $stateProvider.state(customers);
            $stateProvider.state(profile);
            $stateProvider.state(settings);

        }]);

    }
)();
```

My module.run 

```
(
    function()
    {
        'use strict';

        angular.module('Appt.Main',['ngRoute','ui.router','Appt.Directives'])
            .run(['$state','$stateParams', function ($state,$stateParams) {
                console.log('Appt.Main is now running')


                console.log($state.current);
                console.log($stateParams);

                $state.transitionTo('calendar');


            }])
    }
)();
```

**Comment by eddiemonge** [comment]
`$state.current` isn't going to have things in the run block. Do it on the `$stateChangeSuccess` event instead. Im not sure what your server is doing, which are you using?

**Comment by reggieboyYEAH** [comment]
Hello @eddiemonge , I am using express in the backend. 

and here is my code, basically i am returning the same html file for the matching routes

app.get('/:var(/|app/calendar|app/customers|app/profile|app/settings)?', function(req, res) {
    res.sendFile('/app/main/main.html',{ root: '../Appt/public' });
});


i tried implementing $stateChangeSuccess  inside the run block but that event doesnt happen until i call $state.go() inside the run block.

So i ended up with a solution like this

        angular.module('Appt.Main',['ui.router','Appt.Directives'])
            .run(['$rootScope','$state','$stateParams','$location',function ($rootScope,$state,$stateParams,$location) {
                console.log('Appt.Main is now running')

                $rootScope.$state = $state;
                $rootScope.$stateParams = $stateParams;


                //Handle page refresh
                //TODO: possibly change this to if else statement to handle parameters with regex
                var path = $location.path();
                
                switch(path) {
                    case '/app/calendar':
                        $state.go('calendar');
                        break;
                
                    case '/app/customers':
                        $state.go('customers');
                        break;
                }




                $rootScope.$on('$stateChangeStart', function (event, toState) {
                    console.log('$stateChangeStart');
                    console.log(toState);
                });

                $rootScope.$on('$locationChangeSuccess', function(e) {
                    console.log('$locationChangeSuccess');
                });


                $rootScope.$on('$stateChangeSuccess', function(event, toState, toParams, fromState, fromParams){
                    console.log(event);
                    console.log(toState);
                });

            }])

**Comment by eddiemonge** [comment]
So is this resolved then?

**Comment by anurag1140** [comment]
not working in angular 1.5

**Comment by anurag1140** [comment]
angualr1.5

**Comment by anil826** [comment]
Finally I got a way to to solve this issue by server side as it's more like an issue with AngularJs itself I am using 1.5 Angularjs and I got same issue on reload the page.
But after adding below code in my `server.js` file it is save my day ***but it's not a proper solution or not a good way .***

    app.use(function(req, res, next){
      var d = res.status(404);
         if(d){
         	res.sendfile('index.html');
         }
    });

**Comment by manisjos1** [comment]
is this issue fixed ? need similar work around ?!

**Comment by Sreehari94** [comment]
I have used the ui-router in angular js website and deployed through github pages. Is there anyway to solve 
the reloading issue there?

**Comment by xgtxxxx** [comment]
does anyone has the effective solution?

**Comment by uncooloj** [comment]
What server side language are you using?

**Comment by Haroon96** [comment]
I had a similar issue where, on hitting refresh, the state changed to the state I had specified in `$locationProvider.otherwise()`. I noticed that on refresh the URL on the address bar briefly had a `/` appended at its end before the otherwise state loaded. In this case, if I was on the url `app/profile`, it would briefly change to `app/profile/` which wouldn't match with the url specified for that state to `$stateProvider`. I was able to fix this by just adding a `/` at the end of the state url.

If your state is `app/profile`, try changing it to `app/profile/` inside `$stateProvider`. I'm not sure how this would work out for nested states though.
